# Notebook 1 — Data Cleaning
### Crash Risk Intelligence System
**Dataset:** Maryland Crash Reporting — Drivers Data  
**Records:** 215,034 rows × 39 columns  
**Goal:** Clean raw crash data and prepare it for EDA and feature engineering

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Display settings
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)

print("Libraries imported successfully")

Libraries imported successfully


In [2]:
# Load full dataset
df = pd.read_csv('C:\crash_risk_intelligence\data\Crash_Reporting.csv', low_memory=False)

print(f"Rows: {df.shape[0]:,}")
print(f"Columns: {df.shape[1]}")

Rows: 215,034
Columns: 39


In [3]:
# See first 5 rows
df.head()

,Report Number,Local Case Number,Agency Name,ACRS Report Type,Crash Date/Time,Route Type,Road Name,Cross-Street Name,Off-Road Description,Municipality,Related Non-Motorist,Collision Type,Weather,Surface Condition,Light,Traffic Control,Driver Substance Abuse,Non-Motorist Substance Abuse,Person ID,Driver At Fault,Injury Severity,Circumstance,Driver Distracted By,Drivers License State,Vehicle ID,Vehicle Damage Extent,Vehicle First Impact Location,Vehicle Body Type,Vehicle Movement,Vehicle Going Dir,Speed Limit,Driverless Vehicle,Parked Vehicle,Vehicle Year,Vehicle Make,Vehicle Model,Latitude,Longitude,Location
0,MCP3296002G,240018653,MONTGOMERY,Property Damage Crash,04/21/2024 06:53:00 PM,County Route,NaN,NaN,NaN,NaN,NaN,"Sideswipe, Opposite Direction",Clear,Dry,Daylight,No Controls,"Unknown, Unknown",NaN,8F256727-5915-4435-9CE7-5F2D943B8DCA,No,NaN,NaN,NaN,NaN,FDDB5331-EC22-41B2-9143-124692C70CF6,Disabling,Eleven O Clock,Passenger Car,Turning Left,Northbound,25,No,No,2013,HYUNDAI,SONATA,39.105339,-76.989845,"(39.10533874, -76.98984545)"
1,MCP276700BF,240012321,MONTGOMERY,Property Damage Crash,03/14/2024 05:23:00 PM,US (State),COLUMBIA PIKE (SB/L),NaN,NaN,NaN,NaN,"Sideswipe, Same Direction",Clear,Dry,Daylight,No Controls,"Unknown, Unknown",NaN,BB361322-2BD1-47AE-B6B4-9359BAE743A5,No,No Apparent Injury,NaN,NaN,NaN,7EDA4B0A-1880-4BC7-AB71-DBE62C3A9801,Vehicle Not at Scene,Vehicle Not at Scene,Passenger Car,Overtaking/Passing,Southbound,45,No,No,0,NaN,NaN,39.040003,-76.997223,"(39.04000338, -76.99722287)"
2,MCP32790038,240022955,MONTGOMERY,Property Damage Crash,05/15/2024 07:30:00 AM,Maryland (State) Route,VEIRS MILL RD (WB/L),NaN,NaN,NaN,NaN,Front to Rear,Clear,NaN,Daylight,No Controls,"Unknown, Unknown",NaN,E04C495C-1DE3-4AC9-B507-A9B2C207056C,No,NaN,NaN,NaN,NaN,BBAC2C29-9197-40D8-9B22-9156FA82B6F1,Vehicle Not at Scene,Vehicle Not at Scene,Passenger Car,Moving Constant Speed,Northbound,45,No,No,2006,CHEVROLET,COBALT,39.073096,-77.111254,"(39.0730963, -77.1112541)"
3,MCP34000014,240019831,MONTGOMERY,Property Damage Crash,04/28/2024 05:30:00 PM,County Route,HARDING DR,NaN,NaN,NaN,NaN,Front to Rear,Clear,Dry,Daylight,No Controls,"Unknown, Unknown",NaN,D7E62A5D-C733-490D-AC9C-B5EA089C651E,No,NaN,NaN,NaN,NaN,3ACB0037-C9D5-48C5-B410-BE986F22CA86,Functional,One O Clock,Passenger Car,Moving Constant Speed,Southbound,25,No,No,2012,NISSAN,ALTIMA,39.021944,-77.018300,"(39.02194389, -77.01829961)"
4,MCP3341003F,240011829,MONTGOMERY,Injury Crash,03/12/2024 07:30:00 AM,Private Route,NaN,CONNECTICUT AVE,NaN,NaN,NaN,Angle,Clear,Dry,Daylight,No Controls,"Unknown, Unknown",NaN,41AB0CCC-A509-40FF-9BE4-26BB809D02FE,No,No Apparent Injury,NaN,NaN,NaN,785A7E9D-DC33-45B9-B111-7BA14EAC416B,Functional,Twelve O Clock,Pickup,Turning Left,Southbound,35,No,NaN,2015,CHEVROLET,SILVERADO,39.090304,-77.071180,"(39.09030389, -77.07118035)"


In [4]:
# Calculate missing value percentage for each column
missing = (df.isnull().sum() / len(df) * 100).round(1)
missing = missing[missing > 0].sort_values(ascending=False)

print("Columns with missing values:\n")
print(missing)

Columns with missing values:

Non-Motorist Substance Abuse     97.3
Related Non-Motorist             96.7
Municipality                     91.1
Off-Road Description             90.7
Circumstance                     80.5
Cross-Street Name                20.1
Driver Substance Abuse           14.6
Traffic Control                  13.7
Road Name                        11.8
Surface Condition                11.4
Route Type                        9.8
Drivers License State             6.9
Weather                           6.2
Vehicle Going Dir                 4.8
Vehicle Body Type                 1.5
Driver Distracted By              1.5
Injury Severity                   1.4
Vehicle Movement                  0.8
Light                             0.7
Parked Vehicle                    0.7
Vehicle Make                      0.6
Vehicle Model                     0.6
Collision Type                    0.3
Vehicle Damage Extent             0.1
Vehicle First Impact Location     0.1
dtype: float64


## Step 1 — Drop High Null % Columns
Columns with more than 40% missing values are not useful for analysis or modeling.
Dropping 5 columns:
- Non-Motorist Substance Abuse (97.3%)
- Related Non-Motorist (96.7%)
- Municipality (91.1%)
- Off-Road Description (90.7%)
- Circumstance (80.5%)

In [9]:
# Drop columns with more than 40% missing values
cols_to_drop = [
    'Non-Motorist Substance Abuse',
    'Related Non-Motorist',
    'Municipality',
    'Off-Road Description',
    'Circumstance'
]

df.drop(columns=cols_to_drop, inplace=True)

print(f"Columns dropped: {len(cols_to_drop)}")
print(f"New shape: {df.shape}")

KeyError: "['Non-Motorist Substance Abuse', 'Related Non-Motorist', 'Municipality', 'Off-Road Description', 'Circumstance'] not found in axis"

## Step 2 — Fix Inconsistent Casing
Several columns have the same value written in different cases.
For example 'Clear' and 'CLEAR', 'Unknown' and 'UNKNOWN'.
Standardizing all text columns to Title Case.

In [10]:
# Standardize all text/object columns to Title Case
text_cols = df.select_dtypes(include='object').columns

for col in text_cols:
    df[col] = df[col].str.strip().str.title()

# Verify fix on Weather column
print("Weather unique values after fix:")
print(df['Weather'].value_counts())

Weather unique values after fix:
Weather
Clear                                150628
Cloudy                                21068
Raining                               20609
Rain                                   4215
Snow                                   1898
Unknown                                 992
Foggy                                   676
Wintry Mix                              391
Other                                   370
Sleet                                   218
Blowing Snow                            210
Severe Winds                            154
Fog, Smog, Smoke                        100
Freezing Rain Or Freezing Drizzle        79
Severe Crosswinds                        36
Sleet Or Hail                            18
Blowing Sand, Soil, Dirt                 16
Name: count, dtype: int64


## Step 3 — Fix Crash Date/Time Column
'Crash Date/Time' is currently stored as a string/text.
Converting it to proper datetime format so we can 
extract hour, day, month etc. in Feature Engineering.

In [11]:
# Convert Crash Date/Time to datetime
df['Crash Date/Time'] = pd.to_datetime(df['Crash Date/Time'], errors='coerce')

print("Data type after conversion:", df['Crash Date/Time'].dtype)
print("\nSample values:")
print(df['Crash Date/Time'].head())

Data type after conversion: datetime64[ns]

Sample values:
0   2024-04-21 18:53:00
1   2024-03-14 17:23:00
2   2024-05-15 07:30:00
3   2024-04-28 17:30:00
4   2024-03-12 07:30:00
Name: Crash Date/Time, dtype: datetime64[ns]


## Step 4 — Handle Remaining Missing Values
For columns with moderate missing values we use context-aware filling:
- Categorical columns → fill with 'Unknown' (preserves real distribution)
- Numerical columns → fill with median (robust to outliers)
We do NOT use mean for categorical columns as it distorts distributions.

In [12]:
# Fill categorical columns with 'Unknown'
cat_cols = df.select_dtypes(include='object').columns
df[cat_cols] = df[cat_cols].fillna('Unknown')

# Fill numerical columns with median
num_cols = df.select_dtypes(include=['int64', 'float64']).columns
for col in num_cols:
    df[col] = df[col].fillna(df[col].median())

# Verify no missing values remain
remaining_nulls = df.isnull().sum().sum()
print(f"Total missing values remaining: {remaining_nulls}")

Total missing values remaining: 0


## Step 5 — Remove Duplicates
Checking for and removing exact duplicate rows.
Duplicate rows can skew analysis and model training
by giving certain crashes more weight than others.

In [13]:
# Check duplicates before
print(f"Duplicate rows before: {df.duplicated().sum()}")

# Drop duplicates
df.drop_duplicates(inplace=True)

# Check duplicates after
print(f"Duplicate rows after: {df.duplicated().sum()}")
print(f"Final shape: {df.shape}")

Duplicate rows before: 0
Duplicate rows after: 0
Final shape: (215034, 34)


## Step 6 — Fix Mixed Type Columns
'Local Case Number' had mixed types (numbers and strings).
Converting it to string since it is an ID column, not used for math.

In [14]:
# Convert ID-like columns to string
df['Local Case Number'] = df['Local Case Number'].astype(str)

print("Local Case Number dtype:", df['Local Case Number'].dtype)
print("Sample values:")
print(df['Local Case Number'].head())

Local Case Number dtype: object
Sample values:
0    240018653
1    240012321
2    240022955
3    240019831
4    240011829
Name: Local Case Number, dtype: object


## Step 7 — Save Cleaned Data
Saving the cleaned dataframe as a new CSV file for use in the next notebook.
We also save a 1000 row sample for GitHub since the full file is too large to push.

In [15]:
# Save full cleaned data for local use
df.to_csv('C:\crash_risk_intelligence\data\cleaned_crash_data.csv', index=False)
print(f"Full cleaned data saved — shape: {df.shape}")

# Save small sample for GitHub
df.sample(1000, random_state=42).to_csv('C:\crash_risk_intelligence\data\sample_data.csv', index=False)
print("Sample data saved — 1000 rows")

Full cleaned data saved — shape: (215034, 34)
Sample data saved — 1000 rows


## Summary — Data Cleaning Complete
Here is what was done in this notebook:
1. Dropped 5 columns with more than 40% missing values
2. Standardized all text columns to Title Case
3. Converted Crash Date/Time to datetime format
4. Filled remaining missing values appropriately
5. Removed duplicate rows
6. Fixed mixed type columns
7. Saved cleaned data for next notebook

In [16]:
print("=" * 45)
print("       DATA CLEANING COMPLETE")
print("=" * 45)
print(f"Original columns : 39")
print(f"Final columns    : {df.shape[1]}")
print(f"Total rows       : {df.shape[0]:,}")
print(f"Missing values   : {df.isnull().sum().sum()}")
print(f"Duplicate rows   : {df.duplicated().sum()}")
print(f"Date column type : {df['Crash Date/Time'].dtype}")
print("=" * 45)

       DATA CLEANING COMPLETE
Original columns : 39
Final columns    : 34
Total rows       : 215,034
Missing values   : 0
Duplicate rows   : 0
Date column type : datetime64[ns]
